File: OJ/Leetcode/Algorithm/42. Trapping Rain Water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        le = len(height)
        leftmax = [0 for i in range(le)]
        rightmax = [0 for i in range(le)]
        
        maxn = 0
        for i in range(le):
            leftmax[i] = maxn
            maxn = max(maxn,height[i])
            
        maxn = 0
        for i in range(le-1, -1, -1):
            rightmax[i] = maxn
            maxn = max(maxn,height[i])
        
        print height
        print rightmax
        print leftmax
            
        ans = 0
        for i in range(le):
            if leftmax[i] != 0 and rightmax[i] != 0:
                h = min(leftmax[i], rightmax[i]) - height[i]
                if h > 0:
                    ans += h
        return ans
        

File: OJ/Leetcode/Algorithm/42. Trapping Rain Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int ans = 0, n = height.size();
        int max_i = 0, max_n = -1;
        for(int i = 0; i < n; i++)
        {
            if(height[i] > max_n)
            {
                max_i = i;
                max_n = height[i];
            }
        }
        int max_cur = -1;
        for(int i = 0; i < max_i; i++)
        {
            if(max_cur < height[i])
                max_cur = height[i];
            ans += max_cur - height[i];
        }
        max_cur = -1;
        for(int i = n - 1; i > max_i; i--)
        {
            if(max_cur < height[i])
                max_cur = height[i];
            ans += max_cur - height[i];
        }
        return ans;
    }
};
